compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/serialization/KotlinIrLinker.kt

```diff
@@ -190,71 +275,84 @@ abstract class KotlinIrLinker(
         val proto = loadTopLevelDeclarationProto(uniqIdKey)
-        return deserializeDeclaration(proto, reversedFileIndex[uniqIdKey]!!)
+        return deserializersForModules[uniqIdKey.moduleOfOrigin]!!
+            .deserializeDeclaration(proto, reversedFileIndex[uniqIdKey]!!)
     }
+
     protected abstract fun reader(moduleDescriptor: ModuleDescriptor, uniqId: UniqId): ByteArray
 
     private fun loadTopLevelDeclarationProto(uniqIdKey: UniqIdKey): KotlinIr.IrDeclaration {
         val stream = reader(uniqIdKey.moduleOfOrigin!!, uniqIdKey.uniqId).codedInputStream
         return KotlinIr.IrDeclaration.parseFrom(stream, newInstance())
     }
 
-    private fun findDeserializedDeclarationForDescriptor(descriptor: DeclarationDescriptor): DeclarationDescriptor? {
-        val topLevelDescriptor = descriptor.findTopLevelDescriptor()
-
-        // This is Native specific. Try to eliminate.
-        if (topLevelDescriptor.module.isForwardDeclarationModule) return null
-
-        if (topLevelDescriptor !is DeserializedClassDescriptor && topLevelDescriptor !is DeserializedCallableMemberDescriptor) {
-            return null
-        }
-
-        val descriptorUniqId = topLevelDescriptor.getUniqId()
-            ?: error("could not get descriptor uniq id for $topLevelDescriptor")
-        val uniqId = UniqId(descriptorUniqId, isLocal = false)
-        val topLevelKey = UniqIdKey(topLevelDescriptor.module, uniqId)
-
-        // This top level descriptor doesn't have a serialized IR declaration.
-        if (topLevelKey.moduleOfOrigin == null) return null
-
-        reachableTopLevels.add(topLevelKey)
+    private fun deserializeFileAnnotationsIfFirstUse(module: ModuleDescriptor, file: IrFile) {
+        val annotations = fileAnnotations[file] ?: return
+        file.annotations.addAll(deserializersForModules[module]!!.deserializeAnnotations(annotations))
+        fileAnnotations.remove(file)
+    }
 
+    private fun deserializeAllReachableTopLevels() {
         do {
             val key = reachableTopLevels.first()
+            val moduleOfOrigin = key.moduleOfOrigin
 
             if (deserializedSymbols[key]?.isBound == true ||
                 // The key.moduleOrigin is null for uniqIds that we haven't seen in any of the library headers.
                 // Just skip it for now and handle it elsewhere.
-                key.moduleOfOrigin == null) {
+                moduleOfOrigin == null
+            ) {
 
                 reachableTopLevels.remove(key)
                 deserializedTopLevels.add(key)
                 continue
             }
 
-            deserializedModuleDescriptor = key.moduleOfOrigin
             val reachable = deserializeTopLevelDeclaration(key)
             val file = reversedFileIndex[key]!!
             file.declarations.add(reachable)
             reachable.patchDeclarationParents(file)
+            deserializeFileAnnotationsIfFirstUse(moduleOfOrigin, file)
```

Comment:
I think would be good to put changes about file annotations in separate commit.

Reply:
made a separate commit